### PR TITLE
fix: don't panic when logs waits for more than 5 seconds (v2)

### LINKED ISSUE #606: [Bug]: can not get logs from container which is dead or marked for removal
### Testcontainers version

0.15.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host Arch

amd64

### Go Version

1.19

### Docker version

```shell
Client: Docker Engine - Community
 Version:           20.10.21
 API version:       1.41
 Go version:        go1.18.7
 Git commit:        baeda1f
 Built:             Tue Oct 25 18:04:24 2022
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.21
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.18.7
  Git commit:       3056208
  Built:            Tue Oct 25 18:02:38 2022
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.9
  GitCommit:        1c90a442489720eec95342e1789ee8a5e1b9536f
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### Docker info

```shell
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Docker Buildx (Docker Inc., v0.9.1-docker)
  scan: Docker Scan (Docker Inc., v0.21.0)

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 5
 Server Version: 20.10.21
 Storage Driver: overlay2
  Backing Filesystem: xfs
  Supports d_type: true
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 1c90a442489720eec95342e1789ee8a5e1b9536f
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 3.10.0-1160.71.1.el7.x86_64
 Operating System: CentOS Linux 7 (Core)
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 15.51GiB
 Name: dev-bgr-test1-lnx7.iv.local
 ID: JCOB:B6QD:MFHY:4B4Y:Z7OA:H7LA:Z5NM:TDOA:7DD7:AMMB:ZBSJ:4KIU
 Docker Root Dir: /var/lib/docker
 Debug Mode: true
  File Descriptors: 34
  Goroutines: 41
  System Time: 2022-11-02T17:54:20.539106249+01:00
  EventsListeners: 0
```

### What happened?

I'm starting a container and waiting for the end of its execution (it is the expected behavior). I also have a log consumer to display what is going on.

```
	c, err := testcontainers.GenericContainer(context.Background(), testcontainers.GenericContainerRequest{
		ContainerRequest: testcontainers.ContainerRequest{
			Image:    image,
			Networks: []string{network},
		},
	})
	if err != nil {
		return err
	}
	defer c.Terminate(context.Background())

	c.FollowOutput(&stdoutLogConsumer{})

	err = c.Start(context.Background())
	if err != nil {
		return err
	}

	err = c.StartLogProducer(context.Background())
	if err != nil {
		return err
	}

	err = wait.ForExit().WithExitTimeout(180*time.Second).WaitUntilReady(context.Background(), c)
	if err != nil {
		return err
	}

	err = c.StopLogProducer()
	if err != nil {
		return err
	}
```

Sometimes the test fails with this error:

```
     | panic: Error response from daemon: can not get logs from container which is dead or marked for removal
     | goroutine 51 [running]:
     | github.com/testcontainers/testcontainers-go.(*DockerContainer).StartLogProducer.func1()
     | 	/go/pkg/mod/github.com/testcontainers/testcontainers-go@v0.15.0/docker.go:576 +0x5f6
     | created by github.com/testcontainers/testcontainers-go.(*DockerContainer).StartLogProducer
     | 	/go/pkg/mod/github.com/testcontainers/testcontainers-go@v0.15.0/docker.go:558 +0x8a
```

I think there is a problem with the way the log producer is handling a container that terminates: 

```
		r, err := c.provider.client.ContainerLogs(ctx, c.GetContainerID(), options)
		if err != nil {
			// if we can't get the logs, panic, we can't return an error to anything
			// from within this goroutine
			panic(err)
		}
```

I suggest to log a warning and stop log producer instead of calling panic.

### Relevant log output

_No response_

### Additional Information

_No response_

### LINKED ISSUE #946: [Bug]: log producer panics if no logs are produced when using podman
### Testcontainers version

v0.19.0

### Using the latest Testcontainers version?

Yes

### Host OS

Linux

### Host arch

x86_64

### Go version

go1.20

### Docker version

```shell
podman version
Client:       Podman Engine
Version:      4.4.2
API Version:  4.4.2
Go Version:   go1.19.6
Built:        Wed Mar  1 12:22:59 2023
OS/Arch:      linux/amd64
```
```


### Docker info

```shell
podman info:

host:
  arch: amd64
  buildahVersion: 1.29.0
  cgroupControllers:
  - cpu
  - io
  - memory
  - pids
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.7-2.fc37.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.7, commit: '
  cpuUtilization:
    idlePercent: 94.41
    systemPercent: 0.97
    userPercent: 4.62
  cpus: 12
  distribution:
    distribution: fedora
    variant: workstation
    version: "37"
  eventLogger: journald
  hostname: fedora
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 6.1.18-200.fc37.x86_64
  linkmode: dynamic
  logDriver: journald
  memFree: 1141927936
  memTotal: 33228570624
  networkBackend: netavark
  ociRuntime:
    name: crun
    package: crun-1.8.1-1.fc37.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 1.8.1
      commit: f8a096be060b22ccd3d5f3ebe44108517fbf6c30
      rundir: /run/user/1000/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  remoteSocket:
    exists: true
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID
    rootless: true
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.0-8.fc37.x86_64
    version: |-
      slirp4netns version 1.2.0
      commit: 656041d45cfca7a4176f6b7eed9e4fe6c11e8383
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 8272998400
  swapTotal: 8589930496
  uptime: 5h 36m 47.00s (Approximately 0.21 days)
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  volume:
  - local
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - docker.io
  - quay.io
store:
  configFile: /home/martin/.config/containers/storage.conf
  containerStore:
    number: 4
    paused: 0
    running: 4
    stopped: 0
  graphDriverName: overlay
  graphOptions: {}
  graphRoot: /home/martin/.local/share/containers/storage
  graphRootAllocated: 510389125120
  graphRootUsed: 55164096512
  graphStatus:
    Backing Filesystem: btrfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 182
  runRoot: /run/user/1000/containers
  transientStore: false
  volumePath: /home/martin/.local/share/containers/storage/volumes
version:
  APIVersion: 4.4.2
  Built: 1677669779
  BuiltTime: Wed Mar  1 12:22:59 2023
  GitCommit: ""
  GoVersion: go1.19.6
  Os: linux
  OsArch: linux/amd64
  Version: 4.4.2
```

### What happened?

The worker goroutine in DockerContainer.StartLogProducer panics:

```
panic: Get "http://%2Frun%2Fuser%2F1000%2Fpodman%2Fpodman.sock/v1.41/containers/17131d3a1a9090c17437423165cbaf4324140e875b5f69c0ff5c4ee0a78711c5/logs?follow=1&since=1679409204.942642010&stderr=1&stdout=1&tail=": context deadline exceeded

goroutine 66 [running]:
github.com/testcontainers/testcontainers-go.(*DockerContainer).StartLogProducer.func1()
        /home/martin/go/pkg/mod/github.com/testcontainers/testcontainers-go@v0.19.0/docker.go:604 +0x617
created by github.com/testcontainers/testcontainers-go.(*DockerContainer).StartLogProducer
        /home/martin/go/pkg/mod/github.com/testcontainers/testcontainers-go@v0.19.0/docker.go:586 +0x8a

```

This is because:
* there is a harcoded 5 second timeout
* `since` parameter is current time
* the container does not produce log output after the since parameter, there are logs only before that time
* `podman` does not return HTTP response headers until some log output is available

So the context deadline expires and the code panics.

It seems that instead of a hardcoded 5 second timeout, the context should be cancelled when the producer is stopped.

Unless this is a bug in podman, in that case we can open a bug report there.

### Relevant log output

_No response_

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by martin-sucha**
This should fix https://github.com/testcontainers/testcontainers-go/issues/946

## What does this PR do?

There are two commits.

The first commit fixes synchronization issues in producer tests.

The second commit removes panic when logs endpoint takes more than 5 seconds to respond. The panic happened at least with podman when no new logs appear when using follow and since parameters.

We keep retrying until the context is canceled (the retry request would fail anyway) or the producer is stopped, whichever comes first. This makes the retry behavior consistent. The HTTP calls to fetch logs are interrupted immediately (previously the producer waited for in-progress calls to complete).

## Why is it important?

Panicking the whole test suite breaks all my tests if the container ends early or there is some temporary error fetching the logs.

## Related issues

- Closes #946 
- Closes #606 
- Supersedes #947

### CODE REVIEW

File: logconsumer_test.go [hunk 1]
```diff
@@ -31,7 +42,41 @@ func (g *TestLogConsumer) Accept(l Log) {
 		return
 	}
 
-	g.Msgs = append(g.Msgs, s)
+	g.mu.Lock()
```
Comment (mdelapenya) [inline logconsumer_test.go:45]: Could we add a comment here explaining why the synchronisation must be done at the test level? You already added it in #947 :)

**Review by mdelapenya — COMMENTED** [review]
I'm testing this PR with the same script as in #947:

```
for x in {1..20}; do go test -timeout 600s -run ^Test_StartStop$ github.com/testcontainers/testcontainers-go -count=1; done
```

And in general I still see certain flakiness, although different to the one described in that ticket: instead of receiving inconsistent messages, I get a `context deadline exceeded` error. 

Could you take a look at it?

Apart from that, the code at this moment LGTM

**Comment by mdelapenya** [comment]
Hey @martin-sucha I came back from parental leave and I'm revisiting this PR. Did you saw my previous comments?

**Comment by martin-sucha** [comment]
Hey @mdelapenya yes, thank you for the comments! I've been quite busy recently, this keeps falling through the cracks. I hope I can revisit this soon.

**Comment by martin-sucha** [comment]
> I'm testing this PR with the same script as in #947:
> 
> ```
> for x in {1..20}; do go test -timeout 600s -run ^Test_StartStop$ github.com/testcontainers/testcontainers-go -count=1; done
> ```
> 
> And in general I still see certain flakiness, although different to the one described in that ticket: instead of receiving inconsistent messages, I get a `context deadline exceeded` error.
> 

I fixed the error handling so that context canceled is not treated as an error. It is merely a signal that the log producer should stop. Also updated the error handling for other errors to always retry until the log producer is stopped.

**Comment by Tofel** [comment]
wdyt about a different approach:
* allow the caller to pass an optional timeout duration
* instead of panicking return an error if log producer cannot be started?

In this way caller would have bigger control over what it wants to do:
* retry
* pass the error higher up the call stack
* panic, etc

Why?:
* retrying until success might not always be what we want (what if it never succeeds and we retry it until test timeouts?)
* panicking inside a different goroutine makes it impossible to handle it in any way and some programs might be fine with running without logs

**Comment by mdelapenya** [comment]
> wdyt about a different approach:
> 
> * allow the caller to pass an optional timeout duration
> * instead of panicking return an error if log producer cannot be started?
> 
> In this way caller would have bigger control over what it wants to do:
> 
> * retry
> * pass the error higher up the call stack
> * panic, etc
> 
> Why?:
> 
> * retrying until success might not always be what we want (what if it never succeeds and we retry it until test timeouts?)
> * panicking inside a different goroutine makes it impossible to handle it in any way and some programs might be fine with running without logs

@Tofel I like this. Would you be able to work on it? We are thinking about an eventual v1 version, so adding a breaking change in the signature/behaviour, if done properly, wouldn't be that bad.

**Comment by Tofel** [comment]
@mdelapenya I will be able to work on that, yes, but I need to confirm with me team when. Will circle back once I know it.

**Comment by mdelapenya** [comment]
> @mdelapenya it's me again... so my team is okay with me taking the time to propose an alternative solution (although I guess it will have to be a bit different, since we cannot return the error from that goroutine, but why not have an error channel on which the caller could listen in a dedicated goroutine?).

Indeed, I admit that we should totally rewrite the log consumer implementation for the v1, as I think it has grown in a non-organic manner.

> Before I do start with it, could you tell me when you're thinking that v1 might see the light of the day?

Not sure about it. I already have a draft doc with the things that I'd like to do/fix, which I'd make it public right after completed, in the form of a GH discussion. From there, I'll start creating issues in case the community wants to participate.

**Comment by mdelapenya** [comment]
This PR has been superseded by #1971, please take a look at https://github.com/testcontainers/testcontainers-go/pull/1971#issuecomment-1885098616

I'd like to thank you for your time elaborating this PR. It has not been merged this time, but came with a lot of light for the final implementation. Thanks!

**Pull request closed without merging**